kin-ecosystem-sdk/src/main/java/com/kin/ecosystem/data/order/OrderRepository.java

```diff
@@ -157,37 +161,63 @@ private void decrementPendingOrdersCount() {
         if (hasMorePendingOffers()) {
             pendingOrdersCount.decrementAndGet();
         }
-        if (!hasMorePendingOffers()) {
-            blockchainSource.removePaymentObserver(paymentObserver);
-            Log.d(TAG, "decrementPendingOrdersCount: removePaymentObserver");
+        else {
+            if(paymentObserver != null) {
+                blockchainSource.removePaymentObserver(paymentObserver);
+                Log.d(TAG, "decrementPendingOrdersCount: removePaymentObserver");
+            }
         }
     }
 
-    private void setCompletedOrder(Order order) {
+    private void setCompletedOrder(@NonNull Order order) {
         Log.i(TAG, "setCompletedOrder: " + order);
         completedOrder.postValue(order);
         if (!hasMorePendingOffers()) {
-            if (cachedOpenOrder.getValue().getId().equals(order.getOrderId())) {
-                cachedOpenOrder.postValue(null);
-            }
-            if (offerRepository.getPendingOffer().getValue().getId().equals(order.getOfferId())) {
-                offerRepository.setPendingOfferByID(null);
-            }
+            removeCachedOpenOrderByID(order.getOrderId());
+            removePendingOfferByID(order.getOfferId());
         }
     }
 
     private boolean hasMorePendingOffers() {
         return pendingOrdersCount.get() > 0;
     }
 
+    private void removeCachedOpenOrderByID(String orderId) {
+        if (isCachedOpenOrderEquals(orderId)) {
+            cachedOpenOrder.postValue(null);
+        }
+    }
+
+    private void removePendingOfferByID(String offerId) {
+        if (isCurrentPendingOfferIdEquals(offerId)) {
+            offerRepository.setPendingOfferByID(null);
+        }
+    }
+
+    private boolean isCachedOpenOrderEquals(String orderId) {
+        if (cachedOpenOrder != null && cachedOpenOrder.getValue() != null) {
```

Comment:
there is a chance of NPE - cachedOpenOrder.getValue() can be set to null  just after the if check.
You can just get the value into final variable check to avoid such problem

Reply:
👍